src/main/java/com/aws/greengrass/mqttclient/IotCoreTopicValidator.java

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.aws.greengrass.mqttclient;
+
+import com.aws.greengrass.util.Utils;
+import lombok.NonNull;
+
+import java.util.Arrays;
+import java.util.Optional;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+import static com.aws.greengrass.mqttclient.MqttClient.MQTT_VERSION_5;
+
+
+public final class IotCoreTopicValidator {
+
+    public enum Operation {
+        PUBLISH,
+        SUBSCRIBE
+    }
+
+    public static final int TOPIC_MAX_NUMBER_OF_FORWARD_SLASHES = 7;
+    public static final int MAX_LENGTH_OF_TOPIC = 256;
+    private static final int MAX_LENGTH_FOR_UNKNOWN_RESERVED_TOPIC = 512;
+
+    private static final String SHARED_SUBSCRIPTION_TEMPLATE = "^\\$share/\\S+/\\S+";
+    private static final String SHARED_SUBSCRIPTION_PREFIX = "^\\$share/\\S+?/";
+    private static final String DIRECT_INGEST_TEMPLATE = "^\\$aws/rules/\\S+/\\S+";
+    private static final String DIRECT_INGEST_PREFIX = "^\\$aws/rules/\\S+?/";
+
+    private IotCoreTopicValidator() {
+    }
+
+    /**
+     * Check that a given topic adheres to IoT Core limits,
+     * such as number of forward slashes and length.
+     *
+     * @param topic       topic
+     * @param mqttVersion mqtt version (mqtt3, mqtt5)
+     * @param operation   operation
+     * @throws MqttRequestException if the topic is deemed to be invalid
+     */
+    public static void validateTopic(@NonNull String topic,
+                                     @NonNull String mqttVersion,
+                                     @NonNull Operation operation) throws MqttRequestException {
+        if (Utils.isEmpty(topic)) {
+            throw new MqttRequestException("Topic must not be empty");
+        }
+
+        if (operation == Operation.PUBLISH && (topic.contains("#") || topic.contains("+"))) {
+            throw new MqttRequestException("Publish topics must not contain wildcard "
+                    + "characters of '#' or '+'");
+        }
+
+        topic = topic.toLowerCase().trim();
```

Comment:
is this necessary? This will be creating a duplicated temporary string. We use `.intern` on MQTT topics to reduce how much memory we need for topics, so this is kind of undoing that.

Reply:
ah yeah good point, just addressed